File: shared/src/commonTest/kotlin/com/example/tp4/PokerTest.kt
package com.example.tp4

import kotlin.test.Test
import kotlin.test.assertEquals

class PokerTest {
    private val poker = Poker()

    private fun handOf(vararg cards: Card) = Hand(cards.toMutableList())

    @Test
    fun royalFlush() {
        val hand = handOf(
            Card(1, Card.Color.PIQUE),
            Card(13, Card.Color.PIQUE),
            Card(12, Card.Color.PIQUE),
            Card(11, Card.Color.PIQUE),
            Card(10, Card.Color.PIQUE)
        )
        assertEquals(Poker.PokerHandValue.ROYAL_FLUSH, poker.evaluateHand(hand))
    }

    @Test
    fun straightFlush() {
        val hand = handOf(
            Card(2, Card.Color.COEUR),
            Card(3, Card.Color.COEUR),
            Card(4, Card.Color.COEUR),
            Card(5, Card.Color.COEUR),
            Card(6, Card.Color.COEUR)
        )
        assertEquals(Poker.PokerHandValue.STRAIGHT_FLUSH, poker.evaluateHand(hand))
    }

    @Test
    fun fourOfAKind() {
        val hand = handOf(
            Card(7, Card.Color.COEUR),
            Card(7, Card.Color.CARREAU),
            Card(7, Card.Color.PIQUE),
            Card(7, Card.Color.TREFLE),
            Card(2, Card.Color.COEUR)
        )
        assertEquals(Poker.PokerHandValue.FOUR_OF_A_KIND, poker.evaluateHand(hand))
    }

    @Test
    fun fullHouse() {
        val hand = handOf(
            Card(9, Card.Color.COEUR),
            Card(9, Card.Color.CARREAU),
            Card(9, Card.Color.PIQUE),
            Card(4, Card.Color.TREFLE),
            Card(4, Card.Color.COEUR)
        )
        assertEquals(Poker.PokerHandValue.FULL_HOUSE, poker.evaluateHand(hand))
    }

    @Test
    fun aceLowStraight() {
        val hand = handOf(
            Card(1, Card.Color.COEUR),
            Card(2, Card.Color.CARREAU),
            Card(3, Card.Color.PIQUE),
            Card(4, Card.Color.TREFLE),
            Card(5, Card.Color.COEUR)
        )
        assertEquals(Poker.PokerHandValue.STRAIGHT, poker.evaluateHand(hand))
    }

    @Test
    fun richPairJacksOrBetter() {
        val hand = handOf(
            Card(11, Card.Color.COEUR),
            Card(11, Card.Color.CARREAU),
            Card(2, Card.Color.PIQUE),
            Card(4, Card.Color.TREFLE),
            Card(6, Card.Color.COEUR)
        )
        assertEquals(Poker.PokerHandValue.RICH_PAIR, poker.evaluateHand(hand))
    }

    @Test
    fun lowPairIsNothing() {
        val hand = handOf(
            Card(9, Card.Color.COEUR),
            Card(9, Card.Color.CARREAU),
            Card(2, Card.Color.PIQUE),
            Card(4, Card.Color.TREFLE),
            Card(6, Card.Color.COEUR)
        )
        assertEquals(Poker.PokerHandValue.NOTHING, poker.evaluateHand(hand))
    }

    @Test
    fun payoutMultiplierAppliesToBet() {
        val hand = handOf(
            Card(7, Card.Color.COEUR),
            Card(7, Card.Color.CARREAU),
            Card(7, Card.Color.PIQUE),
            Card(7, Card.Color.TREFLE),
            Card(2, Card.Color.COEUR)
        )
        val result = poker.evaluateHand(hand)
        assertEquals(20, result.multiplier)
        assertEquals(100, 5 * result.multiplier)
    }
}

File: web/src/game/Poker.test.ts
import { describe, expect, it } from "vitest";
import { Card, CardColor } from "./Card";
import { Hand } from "./Hand";
import { evaluateHand, POKER_HAND_MULTIPLIER, PokerHandValue } from "./Poker";

const handOf = (...cards: Card[]) => new Hand(cards);

describe("evaluateHand", () => {
  it("royal flush", () => {
    const hand = handOf(
      new Card(1, CardColor.PIQUE),
      new Card(13, CardColor.PIQUE),
      new Card(12, CardColor.PIQUE),
      new Card(11, CardColor.PIQUE),
      new Card(10, CardColor.PIQUE),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.ROYAL_FLUSH);
  });

  it("straight flush", () => {
    const hand = handOf(
      new Card(2, CardColor.COEUR),
      new Card(3, CardColor.COEUR),
      new Card(4, CardColor.COEUR),
      new Card(5, CardColor.COEUR),
      new Card(6, CardColor.COEUR),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.STRAIGHT_FLUSH);
  });

  it("four of a kind", () => {
    const hand = handOf(
      new Card(7, CardColor.COEUR),
      new Card(7, CardColor.CARREAU),
      new Card(7, CardColor.PIQUE),
      new Card(7, CardColor.TREFLE),
      new Card(2, CardColor.COEUR),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.FOUR_OF_A_KIND);
  });

  it("full house", () => {
    const hand = handOf(
      new Card(9, CardColor.COEUR),
      new Card(9, CardColor.CARREAU),
      new Card(9, CardColor.PIQUE),
      new Card(4, CardColor.TREFLE),
      new Card(4, CardColor.COEUR),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.FULL_HOUSE);
  });

  it("ace-low straight", () => {
    const hand = handOf(
      new Card(1, CardColor.COEUR),
      new Card(2, CardColor.CARREAU),
      new Card(3, CardColor.PIQUE),
      new Card(4, CardColor.TREFLE),
      new Card(5, CardColor.COEUR),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.STRAIGHT);
  });

  it("rich pair (jacks or better)", () => {
    const hand = handOf(
      new Card(11, CardColor.COEUR),
      new Card(11, CardColor.CARREAU),
      new Card(2, CardColor.PIQUE),
      new Card(4, CardColor.TREFLE),
      new Card(6, CardColor.COEUR),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.RICH_PAIR);
  });

  it("low pair is nothing", () => {
    const hand = handOf(
      new Card(9, CardColor.COEUR),
      new Card(9, CardColor.CARREAU),
      new Card(2, CardColor.PIQUE),
      new Card(4, CardColor.TREFLE),
      new Card(6, CardColor.COEUR),
    );
    expect(evaluateHand(hand)).toBe(PokerHandValue.NOTHING);
  });

  it("payout multiplier applies to bet", () => {
    const hand = handOf(
      new Card(7, CardColor.COEUR),
      new Card(7, CardColor.CARREAU),
      new Card(7, CardColor.PIQUE),
      new Card(7, CardColor.TREFLE),
      new Card(2, CardColor.COEUR),
    );
    const result = evaluateHand(hand);
    expect(POKER_HAND_MULTIPLIER[result]).toBe(20);
    expect(5 * POKER_HAND_MULTIPLIER[result]).toBe(100);
  });
});
